### PR TITLE
Changes clickcatcher to dynamically scale based on view range rather than making 1 tile for each tile viewed

### LINKED ISSUE #29342: View Range is fucking crashing the server
Setting view range to 10 (in ghost tab as a ghost) then clicking the view range button again causes the screen to go wild on EVERYONEs screens. I have confirmed it several times through testing with Thunder on basil

also setting the view range to 128 crashes dreamseeker

It only started happening in the last ~12 hours or so, it's fair to assume one of the new PRs are the cause of this

### PR DESCRIPTION
**Pull request by silicons**
Scale()s the icon when range is below 16, if it's above 16, scales the icon to 16 and transforms the rest of the way.
fixes #29342

### CODE REVIEW

File: code/_onclick/autoclick.dm [hunk 1]
```diff
@@ -72,6 +72,7 @@
 
 //Please don't roast me too hard
 /client/MouseMove(object,location,control,params)
+	to_chat(world, "<span class='boldnotice'>DEBUG: Client MouseMove with params [params]</span>")
```
Comment (lzimann) [inline code/_onclick/autoclick.dm]: debug

File: code/_onclick/autoclick.dm [hunk 2]
```diff
@@ -84,6 +85,7 @@
 	return
 
 /client/MouseDrag(src_object,atom/over_object,src_location,over_location,src_control,over_control,params)
+	to_chat(world, "<span class='boldnotice'>DEBUG: Client MouseMove with params [params]</span>")
```
Comment (lzimann) [inline code/_onclick/autoclick.dm]: debug

File: code/_onclick/click.dm [hunk 3]
```diff
@@ -432,33 +432,51 @@
...
+	if(view_size_x > 16 || view_size_y > 16)
+		newicon.Scale((16 * 2 + 1) * world.icon_size,(16 * 2 + 1) * world.icon_size)
+		icon = newicon
+		var/tx = view_size_x/16
+		var/ty = view_size_y/16
+		var/matrix/M = new
+		M.Scale(tx, ty)
+		transform = M
+		screen_loc = "CENTER-16,CENTER-16"
+		to_chat(world, "<span class='boldnotice'>DEBUG: Updating greed by transform, with transformation x/y at [tx]/[ty]</span>")
```
Comment (lzimann) [inline code/_onclick/click.dm]: debug

File: code/_onclick/click.dm [hunk 4]
```diff
@@ -432,33 +432,51 @@
...
+		to_chat(world, "<span class='boldnotice'>DEBUG: Updating greed with size [view_size_x]/[view_size_y]! Icon is at [newicon.Width()]/[newicon.Height()] pixels!</span>")
```
Comment (lzimann) [inline code/_onclick/click.dm]: debug

File: code/_onclick/click.dm [hunk 5]
```diff
@@ -432,33 +432,51 @@
...
 	else
-		var/turf/T = screen_loc2turf(screen_loc, get_turf(usr))
+		var/turf/T = params2turf(modifiers["screen-loc"], get_turf(usr))
 		params += "&catcher=1"
 		if(T)
 			T.Click(location, control, params)
 	. = 1
 
+/obj/screen/click_catcher/MouseMove(location,control,params)
+	to_chat(world, "<span class='boldnotice'>DEBUG: Clickcatcher MouseMove with params [params]</span>")
```
Comment (lzimann) [inline code/_onclick/click.dm]: debug

File: code/_onclick/click.dm [hunk 6]
```diff
@@ -432,33 +432,51 @@
...
+	return ..()
+
+/obj/screen/click_catcher/MouseDrag(over_object,src_location,over_location,src_control,over_control,params)
+	to_chat(world, "<span class='boldnotice'>DEBUG: Clickcatcher MouseDrag with params [params]</span>")
```
Comment (lzimann) [inline code/_onclick/click.dm]: debug

**Review by lzimann — DISMISSED** [review]
debug messages

**Comment by MrPerson** [comment]
I would personally prefer to keep it as different screen object rather than use icon procs. Although let it be known that tiling via screen_loc isn't terribly cheap either.

**Comment by AnturK** [comment]
So what's the actual cause here ?

**Comment by AnturK** [comment]
@kevinz000 well ? This is not getting merged without explanation.

**Comment by kevinz000** [comment]
? Cause of what?

**Comment by kevinz000** [comment]
ok so the cause here is when I did it in the beamrifle PR I 1. forgot to make it global 2. Made it retile the clickcatcher by clearing it and regenerating it for every single tile on the player's screen every time the view range changed.
1 caused it to fuck up everyone's view ranges as everyone got their screen objects screwed with, and for higher view ranges their clickcatcher objects get moved off-screen making BYOND force their view range to higher/lower levels to accomodate. 
2 is just a performance hog and apparently a memory leak? Scrolling with mousewheel as a ghost to change view range generated thousands of instances, so that's probably a bad thing.
This addresses both by making it no longer a global variable and instead different for every client so you don't have everyone's view ranges getting screwed when one person scrolls, as well as making it use icon's Scale() proc to resize the icon larger instead of generating more objects, and when it's too large to Scale() properly (IE: 128 view range, the proc will just crash), it'll use transform to enlarge it. Note that mousemove and drag only gets called when your mouse moves on pixel on the icon. Scale will increase the resolution of mousemove detection/keep it the same when used in conjunction with view range, while transform will reduce it, so I used Scale as well as transform when it's over the limit for Scale, so you don't lose too much mousemove detect resolution.

**Comment by optimumtact** [comment]
This is pretty obvious antur

**Comment by optimumtact** [comment]
@MrPerson I fail to see how book keeping a bunch of single tile instances you have to update for every view in the game is somehow more performant than a single icon scale operation when the view range changes?

**Comment by optimumtact** [comment]
Basically it's much less memory intensive to use a whole bunch of clickcatcher icons that map one to one to tiles in the users view. Instead just use a single overlay, however we run into the issue that just transforming the overlay makes the mouse operation precision decrease (as it triggers per pixel and transform just makes the pixels "bigger") so we do a single icon scale operation up to 16 view range.

This shouldn't be a huge issue as the view range does not change very often, so we trade off some instant performance hit to keep the instance count lower.

**Pull request merged**